constellations/orion/src/controllers/classes_by_semesters_controller_test.go

```diff
@@ -0,0 +1,223 @@
+	assert.EqualValues(t, "program1", results[0].ProgramClasses[0].ProgramObj.ProgramId)
+	assert.EqualValues(t, "program1_2020_spring_class1", results[0].ProgramClasses[0].Classes[0].ClassId)
+}
+
+func TestOneSemesterOneProgramOneClass_Failure(t *testing.T) {
+	// Mock 1 program, 1 semester, no classes created
+	testUtils.ProgramRepo.MockSelectAll = func(publishedOnly bool) ([]domains.Program, error) {
+		return createMockPrograms(1), nil
+	}
+	repos.ProgramRepo = &testUtils.ProgramRepo
+
+	testUtils.SemesterRepo.MockSelectAll = func(publishedOnly bool) ([]domains.Semester, error) {
+		return createMockSemesters(1), nil
+	}
+	repos.SemesterRepo = &testUtils.SemesterRepo
+
+	testUtils.ClassRepo.MockSelectAll = func(publishedOnly bool) ([]domains.Class, error) {
+		return []domains.Class{}, errors.New("error in class")
+	}
+	repos.ClassRepo = &testUtils.ClassRepo
+
+	// Create new HTTP request to endpoint
+	recorder := testUtils.SendHttpRequest(t, http.MethodGet, "/api/classesbysemesters", nil)
+
+	// Validate results
+	assert.EqualValues(t, http.StatusNotFound, recorder.Code)
+}
+
+func TestGetClassesAndProgramsBySemester_Success(t *testing.T) {
+	// Mock 2 programs, 2 semesters, 2 classes
+	testUtils.ProgramRepo.MockSelectAll = func(publishedOnly bool) ([]domains.Program, error) {
+		return createMockPrograms(1, 2), nil
+	}
+	repos.ProgramRepo = &testUtils.ProgramRepo
+
+	testUtils.SemesterRepo.MockSelectAll = func(publishedOnly bool) ([]domains.Semester, error) {
+		return createMockSemesters(1, 2), nil
+	}
+	repos.SemesterRepo = &testUtils.SemesterRepo
+
+	testUtils.ClassRepo.MockSelectAll = func(publishedOnly bool) ([]domains.Class, error) {
+		return createMockClasses(1, 2, 3, 4), nil
+	}
+	repos.ClassRepo = &testUtils.ClassRepo
+
+	// Create new HTTP request to endpoint
+	recorder := testUtils.SendHttpRequest(t, http.MethodGet, "/api/classesbysemesters", nil)
+
+	// Validate results
+	assert.EqualValues(t, http.StatusOK, recorder.Code)
+	var results []domains.ProgramClassesBySemester
+	if err := json.Unmarshal(recorder.Body.Bytes(), &results); err != nil {
+		t.Errorf("unexpected error: %v\n", err)
+	}
+	assert.EqualValues(t, "2020_spring", results[0].Semester.SemesterId)
+	assert.EqualValues(t, "program1", results[0].ProgramClasses[0].ProgramObj.ProgramId)
+	assert.EqualValues(t, "program1_2020_spring_class1", results[0].ProgramClasses[0].Classes[0].ClassId)
+	assert.EqualValues(t, "program1_2020_spring_class2", results[0].ProgramClasses[0].Classes[1].ClassId)
+	assert.EqualValues(t, "2020_summer", results[1].Semester.SemesterId)
+	assert.EqualValues(t, "program1", results[1].ProgramClasses[0].ProgramObj.ProgramId)
+	assert.EqualValues(t, "program1_2020_summer_final_review", results[1].ProgramClasses[0].Classes[0].ClassId)
+	assert.EqualValues(t, "program2", results[1].ProgramClasses[1].ProgramObj.ProgramId)
+}
+
+func TestGetOneSemesterTwoProgramsTwoClasses_Success(t *testing.T) {
```

Comment:
OneClass?

Reply:
Mm better yet, maybe name this test names more specifically. Like this test is testing if a program doesn't have any classes. So maybe you can shrink to test JUST this scenario (i.e. 1 semester, 1 program, 0 classes)